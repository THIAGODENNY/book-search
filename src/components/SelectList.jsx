import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import '../styles/components/SelectList.scss';
import InsertItem from './InsertItem';
import CreateList from './CreateList';

const SelectList = ({ selectedOption, onRequestClose }) => {
  const { isOpened, id } = selectedOption;
  const [selectedListItem, setSelectedListItem] = useState();

  const dispatch = useDispatch();
  const {
    data,
    wishList,
    list,
    createListIsOpen,
  } = useSelector((state) => state);
  const [
    setWishlist,
    setSelectedOption,
    setList,
    setCreateListIsOpen,
  ] = [
    (newWishList) => dispatch({ type: 'SET_WISHLIST', wishList: newWishList }),
    (newSelectedOption) => dispatch({ type: 'SET_SELECTED_OPTION', selectedOption: newSelectedOption }),
    (newList) => dispatch({ type: 'SET_LIST', setList: newList }),
    (newCreateListIsOpen) => dispatch({ type: 'SET_CREATE_LIST_IS_OPEN', createListIsOpen: newCreateListIsOpen }),
  ];


  const handleSubmit = (e) => {
    e.preventDefault();

    const { items } = wishList;
    const item = data.items.filter((i) => i.id === id).pop();
    item.listName = e.target.list.value;

    if (items.filter((i) => i.id === id).length === 0) {
      setWishlist({ items: [...wishList.items, item] });
    }
    setSelectedOption({ isOpened: false });
  };

  const handleCreateList = (e) => {
    e.preventDefault();
    setCreateListIsOpen(true);
  };

  const handleCreateListClose = () => {
    setCreateListIsOpen(false);
  };

  const handleSubmitNewList = (e) => {
    e.preventDefault();
    const newListValue = e.target.elements.list.value;
    if (newListValue) {
      setList(newListValue);
    }
    setSelectedListItem(newListValue);
    handleCreateListClose();
  };

  const handleSelectedListChange = (e) => setSelectedListItem(e.target.value);

  return (
    <div>
      <InsertItem
        isOpened={isOpened}
        onRequestClose={onRequestClose}
        handleSubmit={handleSubmit}
        list={list}
        handleCreateList={handleCreateList}
        selectedListItem={selectedListItem}
        handleSelectedListChange={handleSelectedListChange}
      />
      <CreateList
        createListIsOpen={createListIsOpen}
        handleCreateListClose={handleCreateListClose}
        handleSubmitNewList={handleSubmitNewList}
      />
    </div>
  );
};

SelectList.propTypes = {
  selectedOption: PropTypes.arrayOf(
    PropTypes.shape({
      isOpened: PropTypes.bool.isRequired,
      id: PropTypes.string.isRequired,
    }),
  ).isRequired,
  onRequestClose: PropTypes.func.isRequired,
};

export default SelectList;
