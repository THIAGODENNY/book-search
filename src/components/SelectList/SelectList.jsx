import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import InsertItem from '../InsertItem';
import CreateList from '../CreateList';

import * as actionCreator from '../../redux/actions';

const SelectList = ({
  selectedOption,
  onRequestClose,
  data,
  wishList,
  list,
  createListIsOpen,
  selectActions,
}) => {
  const { isOpened, id } = selectedOption;

  const {
    updateList,
    updateSelectedOption,
    updateWishList,
    handleCreateListClose,
    handleCreateList,
  } = selectActions;

  const [selectedListItem, setSelectedListItem] = useState();

  const handleSubmit = (e) => {
    e.preventDefault();

    const { items } = wishList;
    const item = data.items.filter((i) => i.id === id).pop();
    item.listName = e.target.list.value;

    if (items.filter((i) => i.id === id).length === 0) {
      updateWishList({ items: [...wishList.items, item] });
    }
    updateSelectedOption({ isOpened: false });
  };

  const handleSubmitNewList = (e) => {
    e.preventDefault();
    const newListValue = e.target.elements.list.value;
    if (newListValue) {
      updateList(newListValue);
    }
    setSelectedListItem(newListValue);
    handleCreateListClose();
  };

  const handleSelectedListChange = (e) => setSelectedListItem(e.target.value);

  const handleBlockScroll = () => {
    if (isOpened) {
      document.body.style.overflow = 'hidden';
    } else {
      document.body.style.overflow = 'unset';
    }
  };

  useEffect(() => {
    handleBlockScroll();
  }, [isOpened]);

  return (
    <div>
      <InsertItem
        isOpened={isOpened}
        onRequestClose={onRequestClose}
        handleSubmit={handleSubmit}
        list={list}
        handleCreateList={handleCreateList}
        selectedListItem={selectedListItem}
        handleSelectedListChange={handleSelectedListChange}
        id={id}
      />
      <CreateList
        createListIsOpen={createListIsOpen}
        handleCreateListClose={handleCreateListClose}
        handleSubmitNewList={handleSubmitNewList}
      />
    </div>
  );
};

SelectList.propTypes = {
  selectedOption: PropTypes.shape({
    isOpened: PropTypes.bool,
    id: PropTypes.string,
  }).isRequired,
  onRequestClose: PropTypes.func.isRequired,
  data: PropTypes.shape({
    items: PropTypes.array,
  }).isRequired,
  wishList: PropTypes.shape({
    items: PropTypes.array,
  }).isRequired,
  list: PropTypes.arrayOf(PropTypes.string).isRequired,
  createListIsOpen: PropTypes.bool.isRequired,
  selectActions: PropTypes.shape({
    updateList: PropTypes.func.isRequired,
    updateSelectedOption: PropTypes.func.isRequired,
    updateWishList: PropTypes.func.isRequired,
    handleCreateListClose: PropTypes.func.isRequired,
    handleCreateList: PropTypes.func.isRequired,
  }).isRequired,
};

SelectList.defaults = {
  selectedOption: {
    isOpened: false,
    id: '',
  },
};

const mapStateToProps = (state) => ({
  data: state.data,
  wishList: state.wishList,
  list: state.list,
  createListIsOpen: state.createListIsOpen,
});

const mapDispatchToProps = (dispatch) => ({
  selectActions: bindActionCreators(actionCreator, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(SelectList);
