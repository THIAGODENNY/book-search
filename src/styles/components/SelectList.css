.modal {
  position: absolute;
  display: flex;
  align-items: center;
  top: 35%;
  left: 15%;
  right: 15%;
  bottom: 35%;
  min-height: 100px;
  min-width: 350px;
  height: fit-content;
  width: fit-content;
  background-color: white;
}

.submit {
  width: auto;
}

.submit-form, .submit-form-new-list {
  display: flex;
  flex-direction: row;
  justify-content: center;
  width: 100%;
}