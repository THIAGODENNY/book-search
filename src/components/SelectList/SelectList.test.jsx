import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import { storeFactory } from '../../tests/storeFactory';
import SelectList from './SelectList';
import '@testing-library/jest-dom';

const initialState = (props) => ({
  data: {
    items: [],
  },
  page: 0,
  wishList: {
    items: JSON.parse(localStorage.getItem('items') || '[]'),
  },
  search: '',
  list: [],
  selectedOption: {
    isOpen: false,
    id: '',
  },
  createListIsOpen: false,
  filter: undefined,
  hasMoreItems: true,
  ...props,
});

describe('testing if components renders correcly', () => {
  it('renders correcly', () => {
    const randomText = 'fooBar';
    const store = storeFactory(initialState({
      data: {
        items: [
          {
            id: '0',
            volumeInfo: {
              title: `title ${randomText}`,
              authors: `authors ${randomText}`,
              description: `description ${randomText}`,
              imageLinks: {
                smallThumbnail: `href ${randomText}`,
              },
              categories: `categories ${randomText}`,
              pageCount: `pages ${randomText}`,
              averageRating: `averageRating ${randomText}`,
              language: `language ${randomText}`,
            },
          },
        ],
      },
      createListIsOpen: true,
    }));
    const { baseElement } = render(
      <Provider store={store}>
        <SelectList
          selectedOption={{ id: '0', isOpened: true }}
          onRequestClose={() => {}}
          createListIsOpen
        />
      </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('tests if components renders correctly', () => {
    const randomText = `${Math.random()}`;
    const store = storeFactory(initialState({
      data: {
        items: [
          {
            id: '0',
            volumeInfo: {
              title: `title ${randomText}`,
              authors: `authors ${randomText}`,
              description: `description ${randomText}`,
              imageLinks: {
                smallThumbnail: `href ${randomText}`,
              },
              categories: `categories ${randomText}`,
              pageCount: `pages ${randomText}`,
              averageRating: `averageRating ${randomText}`,
              language: `language ${randomText}`,
            },
          },
        ],
      },
      createListIsOpen: true,
    }));
    const { getByTestId } = render(
      <Provider store={store}>
        <SelectList
          selectedOption={{ id: '0', isOpened: true }}
          onRequestClose={() => {}}
          createListIsOpen
        />
      </Provider>,
    );

    expect(getByTestId('insert-item__body')).toBeInTheDocument();
    expect(getByTestId('create-list__submit-form')).toBeInTheDocument();
  });

  it('changes list on InsertList when submits on CreateList', () => {
    const randomText = `${Math.random()}`;
    const store = storeFactory(initialState({
      data: {
        items: [
          {
            id: '0',
            volumeInfo: {
              title: `title ${randomText}`,
              authors: `authors ${randomText}`,
              description: `description ${randomText}`,
              imageLinks: {
                smallThumbnail: `href ${randomText}`,
              },
              categories: `categories ${randomText}`,
              pageCount: `pages ${randomText}`,
              averageRating: `averageRating ${randomText}`,
              language: `language ${randomText}`,
            },
          },
        ],
      },
      createListIsOpen: true,
    }));
    const { getByTestId } = render(
      <Provider store={store}>
        <SelectList
          selectedOption={{ id: '0', isOpened: true }}
          onRequestClose={() => {}}
          createListIsOpen
        />
      </Provider>,
    );
    fireEvent.change(getByTestId('create-list__submit-form__input'), { target: { value: `list ${randomText}` } });
    fireEvent.submit(getByTestId('create-list__submit-form'));
    expect(getByTestId('insert-item__body__submit-form__list')).toHaveTextContent(`list ${randomText}`);
  });

  it('blocks scroll when modal is opened', () => {
    const store = storeFactory();
    const component = (isOpened) => (
      <Provider store={store}>
        <SelectList
          selectedOption={{ id: '0', isOpened }}
          onRequestClose={() => {}}
          createListIsOpen
        />
      </Provider>
    );
    const { rerender } = render(
      component(false),
    );

    expect(document.body.style.overflow).toBe('unset');

    rerender(component(true));

    expect(document.body.style.overflow).toBe('hidden');
  });
});
