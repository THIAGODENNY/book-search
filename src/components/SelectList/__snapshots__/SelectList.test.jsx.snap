// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing if components renders correcly renders correcly 1`] = `
<body
  class="ReactModal__Body--open"
  style="overflow: hidden;"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div>
    <div />
  </div>
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        aria-label="Selected Option"
        class="ReactModal__Content ReactModal__Content--after-open insert-item"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="insert-item__body"
          data-testid="insert-item__body"
        >
          <div
            class="insert-item__body__description"
          >
            <h1
              class="insert-item__body__description__title"
            >
              title fooBar
            </h1>
            <span
              class="insert-item__body__description__authors"
            >
              authors fooBar
            </span>
            <span
              class="insert-item__body__description__description"
            >
              description fooBar
            </span>
          </div>
          <input
            alt="logo"
            class="insert-item__body__image"
            src="href fooBar"
            type="image"
          />
          <div
            class="insert-item__body__info"
          >
            <span
              class="insert-item__body__info__categories"
            >
              categories fooBar
            </span>
            <span
              class="insert-item__body__info__pages"
            >
              Pages:
              pages fooBar
            </span>
            <span
              class="insert-item__body__info__rating"
            >
              Rating:
              averageRating fooBar
            </span>
            <span
              class="insert-item__body__info__language"
            >
              Language:
              language fooBar
            </span>
          </div>
          <form
            class="insert-item__body__submit-form"
          >
            <span
              class="insert-item__body__submit-form__title"
            >
              Add to list:
            </span>
            <select
              class="insert-item__body__submit-form__list"
              data-testid="insert-item__body__submit-form__list"
              id="list"
              name="list"
              required=""
            />
            <button
              class="insert-item__body__submit-form__create-list"
              type="button"
            >
              +
            </button>
            <input
              class="insert-item__body__submit-form__submit"
              type="submit"
            />
          </form>
        </div>
      </div>
    </div>
  </div>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open"
      style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
    >
      <div
        class="ReactModal__Content ReactModal__Content--after-open create-list"
        role="dialog"
        tabindex="-1"
      >
        <form
          class="create-list__submit-form"
          data-testid="create-list__submit-form"
        >
          <h1>
            Write a list to create:
          </h1>
          <input
            data-testid="create-list__submit-form__input"
            name="list"
            required=""
            type="text"
          />
          <input
            class="create-list__submit-form__submit"
            type="submit"
          />
        </form>
      </div>
    </div>
  </div>
</body>
`;
